.gitlab-ci.yml

```diff
@@ -0,0 +1,198 @@
+        GIT_STRATEGY: none
+    script:
+        - rm -rf ${BUILD_ROOT}/${PLAT}
+        - rm -f ${BUILD_ROOT}/mfem
+
+# This section is LLNL-specific
+.setup:
+    stage: setup
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - mkdir -p ${BUILD_ROOT} && cd ${BUILD_ROOT}
+        - if [ ! -d "tpls" ]; then git clone ssh://git@mybitbucket.llnl.gov:7999/mfem/tpls.git; fi
+        - if [ ! -d "tests" ]; then git clone ssh://git@mybitbucket.llnl.gov:7999/mfem/tests.git; fi
+        - cd tpls && git pull && cd ..
+        - cd tests && git pull && cd ..
+
+.with_gcc_6_1_0:
+    variables:
+        TOOLCHAIN: gcc_6_1_0
+    before_script:
+        - module load gcc/6.1.0
+
+.with_gcc_4_9_3:
+    variables:
+        TOOLCHAIN: gcc_4_9_3
+    before_script:
+        - module load gcc/4.9.3
+
+.build_hypre:
+    variables:
+        EXEC_PREFIX: ""
+    stage: libs
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - cd ${BUILD_ROOT}/tpls
+        - mkdir -p ${BUILD_PATH}
+        - make clean
+        - |
+          if [[ "${MPIEXEC}" == "srun" ]]
+          then
+              export JOB_ID=$(squeue -h --name=${MFEM_ALLOC_NAME} --format=%A)
+              export EXEC_PREFIX="srun --jobid=${JOB_ID}"
+          fi
+        - $EXEC_PREFIX make -j 4 hypre PREFIX=${BUILD_PATH}/hypre/src/ BUILD_DIR=${BUILD_PATH} LOG_DIR=${BUILD_PATH}
+
+.build_metis:
+    variables:
+        EXEC_PREFIX: ""
+    stage: libs
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - cd ${BUILD_ROOT}/tpls
+        - mkdir -p ${BUILD_PATH}
+        - |
+          if [[ "${MPIEXEC}" == "srun" ]]
+          then
+              export JOB_ID=$(squeue -h --name=${MFEM_ALLOC_NAME} --format=%A)
+              export EXEC_PREFIX="srun --jobid=${JOB_ID}"
+          fi
+        - $EXEC_PREFIX make -j 4 ${METIS} PREFIX=${BUILD_PATH}/ BUILD_DIR=${BUILD_PATH} LOG_DIR=${BUILD_PATH} CC=cc CXX=c++
+
+.build_mfem:
+    variables:
+        EXEC_PREFIX: ""
+    stage: build
+    script:
+        - mkdir -p ${BUILD_PATH}
+        - cd ..
+        - cp -r mfem ${BUILD_PATH}/mfem_${MFEM_USE_MPI}_${MFEM_DEBUG}
+        - cd ${BUILD_PATH}/mfem_${MFEM_USE_MPI}_${MFEM_DEBUG}
+        - |
+          if [[ "${MPIEXEC}" == "srun" ]]
+          then
+              export JOB_ID=$(squeue -h --name=${MFEM_ALLOC_NAME} --format=%A)
+              export EXEC_PREFIX="srun --jobid=${JOB_ID}"
+          fi
+        - make config CXX=c++ MPICXX=mpicxx MFEM_USE_MPI=${MFEM_USE_MPI} MFEM_DEBUG=${MFEM_DEBUG} MFEM_MPIEXEC="$EXEC_PREFIX" MFEM_MPIEXEC_NP=$MPIEXEC_NP
+        - $EXEC_PREFIX make -j 8 all
+
+.sanitycheck_mfem:
+    stage: test
+    variables:
+        GIT_STRATEGY: none
+    script:
+        - cd ${BUILD_PATH}/mfem_${MFEM_USE_MPI}_${MFEM_DEBUG}
+        - make test
+
+# The BUILD_ROOT is unique to the branch and the pipeline ID. Branch is not
+# necessary but makes it easier to retrieve a data. This means that there cannot
+# be concurrency between 2 pipelines.
```

Comment:
I don't understand what is meant by the second sentence above.

Reply:
I am using 2 parameters to name the directory `mfem_${CI_COMMIT_REF_SLUG}_${CI_PIPELINE_ID}`:
- CI_COMMIT_REF_SLUG is the ref name (converted), usually the branch/tag name
- CI_PIPELINE_ID
Since the pipeline ID is unique, ref name is not necessary, but it make the directory name more human readable: I can easily tell where the working dir for my branch is.